docs_nnx/guides/linen_to_nnx.rst

```diff
@@ -0,0 +1,732 @@
+
+
+Training Step and Compilation
+==========
+
+Now we write a training step and compile it using JAX just-in-time compilation. Note a few differences here:
+
+* Linen uses ``@jax.jit`` to compile the training step, whereas NNX uses ``@nnx.jit``.  ``jax.jit`` only accepts pure stateless arguments, but ``nnx.jit`` allows the arguments to be stateful NNX modules. This greatly reduced the number of lines needed for a train step.
+
+* Similarly, Linen uses ``jax.grad()`` to return a raw dictionary of gradients, wheras NNX can use ``nnx.grad`` to return an NNX model state of gradients. In NNX, here you need to use the split/merge API
+
+  * If you are already using Optax optimizer classes like ``optax.adam(1e-3)`` and use its ``update()`` method to update your model params (instead of the raw ``jax.tree.map`` computation like here), check out `nnx.Optimizer example <https://flax-nnx.readthedocs.io/en/latest/nnx_basics.html#transforms>`__ for a much more concise way of training and updating your model.
+
+* The Linen train step needs to return a tree of parameters, as the input of the next step. On the other hand, NNX's step doesn't need to return anything, because the ``model`` was already in-place-updated within ``nnx.jit``.
+
+* NNX modules are stateful and automatically tracks a few things within, such as RNG keys and BatchNorm stats. That's why you don't need to explicitly pass an RNG key in on every step. Note that you can use `nnx.reseed <https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/rnglib.html#flax.nnx.reseed>`__ to reset its underlying RNG state.
+
+* In Linen, you need to explicitly define and pass in an argument ``training`` to control the behavior of ``nn.Dropout`` (namely, its ``deterministic`` flag, which means random dropout only happens if ``training=True``). In NNX, you can call ``model.train()`` to automatically switch ``nnx.Dropout`` to training mode. Conversely, call ``model.eval()`` to turn off training mode. You can learn more about what this API does at its `API reference <https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/module.html#flax.nnx.Module.train>`__.
+
+
+.. codediff::
+  :title: Linen, NNX
+  :sync:
+
+  ...
+
+  @jax.jit
+  def train_step(key, params, inputs, labels):
+    def loss_fn(params):
+      logits = model.apply(
+        {'params': params},
+        inputs, training=True, # <== inputs
+        rngs={'dropout': key}
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+
+    grads = jax.grad(loss_fn)(params)
+
+    params = jax.tree.map(lambda p, g: p - 0.1 * g, params, grads)
+    return params
+
+  ---
+
+  model.train() # Sets ``deterministic=False` under the hood for nnx.Dropout
+
+  @nnx.jit
+  def train_step(model, inputs, labels):
+    def loss_fn(model):
+      logits = model(
+
+        inputs, # <== inputs
+
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
```

Comment:
Could be more appealing to show it like this to contrast how much simpler NNX is in this case.
```suggestion
    def loss_fn(model):
      logits = model(inputs)




      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
```